solidity/contracts/AbstractBondedECDSAKeep.sol

```diff
@@ -0,0 +1,643 @@
+        require(!isInitialized, "Contract already initialized");
+
+        owner = _owner;
+        members = _members;
+        honestThreshold = _honestThreshold;
+
+        status = Status.Active;
+        isInitialized = true;
+
+        /* solium-disable-next-line security/no-block-members*/
+        keyGenerationStartTimestamp = block.timestamp;
+
+        bonding = IBondingManagement(_bonding);
+    }
+
+    /// @notice Checks if the member already submitted a public key.
+    /// @param _member Address of the member.
+    /// @return True if member already submitted a public key, else false.
+    function hasMemberSubmittedPublicKey(address _member)
+        internal
+        view
+        returns (bool)
+    {
+        return submittedPublicKeys[_member].length != 0;
+    }
+
+    /// @notice Returns true if signing of a digest is currently in progress.
+    function isSigningInProgress() internal view returns (bool) {
+        return signingStartTimestamp != 0;
+    }
+
+    /// @notice Marks the keep as closed.
+    /// Keep can be marked as closed only when there is no signing in progress
+    /// or the requested signing process has timed out.
+    function markAsClosed() internal {
+        status = Status.Closed;
+        emit KeepClosed();
+    }
+
+    /// @notice Marks the keep as terminated.
+    /// Keep can be marked as terminated only when there is no signing in progress
+    /// or the requested signing process has timed out.
+    function markAsTerminated() internal {
+        status = Status.Terminated;
+        emit KeepTerminated();
+    }
+
+    /// @notice Coverts a public key to an ethereum address.
+    /// @param _publicKey Public key provided as 64-bytes concatenation of
+    /// X and Y coordinates (32-bytes each).
+    /// @return Ethereum address.
+    function publicKeyToAddress(bytes memory _publicKey)
+        internal
+        pure
+        returns (address)
+    {
+        // We hash the public key and then truncate last 20 bytes of the digest
+        // which is the ethereum address.
+        return address(uint160(uint256(keccak256(_publicKey))));
+    }
+
+    /// @notice Returns bonds to the keep members.
+    function freeMembersBonds() internal {
+        for (uint256 i = 0; i < members.length; i++) {
+            bonding.freeBond(members[i], uint256(address(this)));
+        }
+    }
+
+    /// @notice Punishes keep members after proving a signature fraud.
+    function slashForSignatureFraud() internal;
+
+    /// @notice Gets the beneficiary for the specified member address.
+    /// @param _member Member address.
+    /// @return Beneficiary address.
+    function beneficiaryOf(address _member)
+        internal
+        view
+        returns (address payable);
+
+    /// @notice Terminates the keep.
+    /// Keep can be marked as terminated only when there is no signing in progress
+    /// or the requested signing process has timed out.
+    function terminateKeep() internal;
```

Comment:
`terminateKeep` is declared here but implemented in `BondedECDSAKeep`. The implementation in `BondedECDSAKeep` calls back `markAsTerminated` defined here. We have a similar situation for `closeKeep`. I find it an antipattern. I think we should rather override function in inheriting class and add additional behavior there, where needed.

I propose to implement both `terminateKeep` and `closeKeep` here:
```
    function terminateKeep() internal {
        markAsTerminated();
    }

    function closeKeep() public onlyOwner onlyWhenActive {
        markAsClosed();
        freeMembersBonds();
    }
```

and add the required stake logic specific to KEEP version of keep in `BondedECDSAKeep`:
```
    function terminateKeep() internal {
        super.terminateKeep();
        unlockMemberStakes();
    }

    function closeKeep() public onlyOwner onlyWhenActive {
        super.closeKeep();
        unlockMemberStakes();
    }
```